### PR TITLE
52 Database Feature

### LINKED ISSUE #52: Database Feature
Some working prototype of the firestore database

### PR DESCRIPTION
**Pull request by Jameszf**
Added abstract classes for making DB handlers. 
Added device login and automatic account creation (will replace with registration fragment later).

DBConnector will eventually be removed.
MainActivityDB may or may not be removed in the future.
No test were made for PostEventDB or MainActivityDB.
UML updated.
CRC updated.

image

### CODE REVIEW

File: app/src/main/java/com/example/pygmyhippo/common/Account.java
Comment (SuperCreativeUniqueUsername) [inline app/src/main/java/com/example/pygmyhippo/common/Account.java:1]: why did you make account parcelable but not event

File: app/src/main/java/com/example/pygmyhippo/common/Event.java
Comment (SuperCreativeUniqueUsername) [inline app/src/main/java/com/example/pygmyhippo/common/Event.java:1]: Is it better to pass an event around like you're doing for account, or pass an id between fragments and query from the database.

File: app/src/main/java/com/example/pygmyhippo/database/DBOnCompleteListener.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package com.example.pygmyhippo.database;
+
+import androidx.annotation.NonNull;
+
+import java.util.ArrayList;
+
+/**
+ * OnCompleteListener interface for database queries.
```
Comment (SuperCreativeUniqueUsername) [inline app/src/main/java/com/example/pygmyhippo/database/DBOnCompleteListener.java:8]: i like the example 👍

File: app/src/main/java/com/example/pygmyhippo/organizer/PostEventDB.java
Comment (SuperCreativeUniqueUsername) [inline app/src/main/java/com/example/pygmyhippo/organizer/PostEventDB.java:1]: what is with this, does this work?

File: app/src/main/java/com/example/pygmyhippo/organizer/PostEventFragment.java [hunk 2]
```diff
@@ -99,13 +102,19 @@ public void onClick(View v) {
...
+                    }
+
                     myEvent.setEventLimitCount(eventLimit.isEmpty() ? -1 : Integer.valueOf(eventLimit));
                     myEvent.setEventWinnersCount(Integer.valueOf(eventWinners));
                     myEvent.setEntrants(new ArrayList<>()); // no entrants of a newly created event
                     myEvent.setGeolocation(eventGeolocaion);
                     myEvent.setEventStatus(Event.EventStatus.ongoing); // default is ongoing
-                    // TODO: connect to database
+
+                    dbHandler.addEvent(myEvent, null); // TODO: Get organiser ID and pass it to addEvent.
```
Comment (SuperCreativeUniqueUsername) [inline app/src/main/java/com/example/pygmyhippo/organizer/PostEventFragment.java:124]: does this work?

File: app/src/main/res/layout/loading_fragment.xml
Comment (SuperCreativeUniqueUsername) [inline app/src/main/res/layout/loading_fragment.xml:1]: whats the loading fragment?

Comment (Jameszf) [reply app/src/main/res/layout/loading_fragment.xml:1]: Its for something to display when DB request are loading. It is not used right now but we will have to use it in the future.

File: app/src/main/java/com/example/pygmyhippo/organizer/PostEventDB.java
Comment (Jameszf) [reply app/src/main/java/com/example/pygmyhippo/organizer/PostEventDB.java:1]: yes

File: app/src/main/java/com/example/pygmyhippo/common/Account.java
Comment (Jameszf) [reply app/src/main/java/com/example/pygmyhippo/common/Account.java:1]: We are passing account as arguments around to the fragments in navigation. So Account has to parcelable (as well as facility) but Event does not at the moment (maybe in the future will make it parcelable too).

File: app/src/main/java/com/example/pygmyhippo/common/Event.java
Comment (Jameszf) [reply app/src/main/java/com/example/pygmyhippo/common/Event.java:1]: No idea. I will think about that

File: app/src/main/java/com/example/pygmyhippo/database/DBOnCompleteListener.java (hunk 1 above)
Comment (Jameszf) [reply app/src/main/java/com/example/pygmyhippo/database/DBOnCompleteListener.java:8]: Thanks

**Review by SuperCreativeUniqueUsername — APPROVED** [review]
Looks good to me!

**Comment by ghost** [comment]
I've been receiving this error when I try to compile the branch 
Screenshot 2024-11-02 at 4 44 27 PM

**Comment by Jameszf** [comment]
Jen. I think youre compiling a different branch. The lines that are causing that error are commented out in 52-database-feature.

**Comment by ghost** [comment]
> Jen. I think youre compiling a different branch. The lines that are causing that error are commented out in 52-database-feature.

I double checked the branch, but I think android studio didn't update the code properly when I pulled the branch. It wouldn't be the first time Lol.

**Review by ghost — APPROVED** [review]
Seems good to me, everything is functional

**Review by Kori-Kozicki — APPROVED** [review]
I think all of this looks really good! The use of the flags seems really handy and I like that you made examples on how to extend it.

File: app/src/androidTest/java/com/example/pygmyhippo/user/SingleEventTest.java [hunk 3]
```diff
@@ -36,7 +36,7 @@ public class SingleEventTest {
     // click onto the qr code from home
     @Before
     public void setUp() {
-        onView(withId(R.id.u_scanQRcodeFragment)).perform(click());
+        onView(withId(R.id.u_scanQR_menu_item)).perform(click());
```
Comment (katzhong) [inline app/src/androidTest/java/com/example/pygmyhippo/user/SingleEventTest.java:39]: thanks for renaming this 🤑

File: app/src/main/java/com/example/pygmyhippo/MainActivity.java [hunk 4]
```diff
@@ -159,26 +159,131 @@ public void onRoleSelected(String role) {
...
      * @author Griffin
      * @param navView: not sure
+     * @param bundle: Arguments to pass to first fragment.
      */
-    private void setupNavControllerUser(BottomNavigationView navView) {
+    private void setupNavControllerUser(BottomNavigationView navView, Bundle bundle) {
         AppBarConfiguration appBarConfiguration = new AppBarConfiguration.Builder(
                 // TODO: change the navigation names for user to be more descriptive
-                R.id.navigation_home, R.id.u_scanQRcodeFragment, R.id.navigation_notifications)
+                R.id.u_my_events_menu_item, R.id.u_scanQR_menu_item, R.id.u_profile_menu_item)
```
Comment (katzhong) [inline app/src/main/java/com/example/pygmyhippo/MainActivity.java:167]: love the naming 😍

**Pull request merged**